### DISCUSSION TITLE
Is it possible the generation of synthetic data for clustering problems?

### DISCUSSION THREAD

**Discussion by TxusLopez** [online-ml/river#943] [Q&A] [answered]
Hi,

In some ocassions, I have used the `synth.ConceptDriftStream` for creating my own datasets with concept drift for classification purposes. This time, I need the same but for clustering problems. Checking the `river.datasets.base.SyntheticDataset`, I see in the parameter `task` that only regression and classification tasks are considered. 

I was wondering: 

- whether there were ways in which I could use river tools to create synthetic datasets for clustering
- if that is not possible, is there any way to make compatible the sklearn datasets generated with e.g. `make_blobs` with river, e.g. by using `river.datasets.base.SyntheticDataset`?

Thanks in advance!

**Comment by MaxHalford** [comment, marked as answer]
> Checking the river.datasets.base.SyntheticDataset, I see in the parameter task that only regression and classification tasks are considered.

That's true. Although in theory clustering is agnostic of the nature of the task, so you be able to do clustering on every dataset.

> whether there were ways in which I could use river tools to create synthetic datasets for clustering

If you loop over the dataset, use `x`, and ignore `y`, doesn't that work for you? What problem have you encountered?

> if that is not possible, is there any way to make compatible the sklearn datasets generated with e.g. make_blobs with river, e.g. by using river.datasets.base.SyntheticDataset?

Yes, there's a [`stream.iter_sklearn_dataset`](https://riverml.xyz/0.11.0/api/stream/iter-sklearn-dataset/) function that should suit your needs.

**Reply by TxusLopez** [reply]
Thanks for the prompt reply @MaxHalford 

The matter behind my question is that I have been thinking about creating synthetic datasets with concept drift through `ConceptDriftStream`, where I have to provide two streams as arguments, and thsese streams need to be `river.datasets.base.SyntheticDataset`. So I have not found the way till now.

If you suggest me to create the datasets as for classification problems, and later forget the label and treat them as clustering problems ... it would work for me.

Thanks!

**Reply by MaxHalford** [reply]
> If you suggest me to create the datasets as for classification problems, and later forget the label and treat them as clustering problems ... it would work for me.

I think that's the way to go! But maybe there's some work to do to make this smoother. Let me know how it goes.

**Reply by TxusLopez** [reply]
Seems that the induction of drift for (truly) unsupervised clustering problems is not so straightforward. This means that we should not reuse synthetic datasets with artificial drifts elaborated for classification problems. That is because, without labels in a real unsupervised scenario, we should only focus on feature drifts (the relations between features and classes cannot be considered, they are supposed to be unknown). Then, to artificially create such kind of drifts, e.g. we should select a subset of features and randomly rotating their values for all examples after a given changepoint (this is just the most simple approach, others more refinated exist). And this functionality cannot be covered by the current tools of river for synthetic data generation with drifts.

This is just a first idea on this topic. I hope it sounds interesting.

Thanks a lot for your help @MaxHalford, river is a great tool and it is being constantly improved and refined ... amazing.

**Reply by MaxHalford** [reply]
I think that what you're saying is interesting. To be honest I never looked very deeply into our synthetic datasets. They got ported from scikit-multiflow. My naive thinking was that you could at least use `datasets.synth.ConceptDriftStream` to generate some drift. But ys maybe it's not ideal for clustering problems.

I am aware of [stream-learn](https://github.com/w4k2/stream-learn), which has some [great docs](https://stream-learn.readthedocs.io/en/latest/). I suggested them to merge with us, but it didn't work out.

Let us know if you make any findings, and feel welcome to contribute. I'd be glad to help you out.
